lib/cdo/sauce_connect.rb

```diff
@@ -0,0 +1,113 @@
+        raise "saucelabs_authkey must be set in locals.yml, use 'Access Key' from https://app.saucelabs.com/user-settings" unless CDO.saucelabs_authkey
+        raise "saucelabs_tunnel_name must be set in locals.yml, 'cdo-tunnel' is a good default choice" unless CDO.saucelabs_tunnel_name
+
+        # Regexes defining the localhost domains we want to tunnel:
+        tunnel_domains = [
+          %q(.*\.code.org),
+          %q(.*\.csedweek.org),
+          %q(.*\.hourofcode.com),
+          %q(.*\.codeprojects.org),
+        ].join(',')
+
+        cmd = [
+          "sc", "run",
+          "--region", "us-west-1",
+          "--tunnel-domains", tunnel_domains,
+          "--tunnel-name", CDO.saucelabs_tunnel_name,
+        ]
+
+        env = {
+          "SAUCE_USERNAME" => CDO.saucelabs_username,
+          "SAUCE_ACCESS_KEY" => CDO.saucelabs_authkey
+        }
+
+        pid = Process.spawn(env, *cmd, out: log_file.path, err: log_file.path)
+        log "starting sc, pid = #{pid}, see log at #{log_file.path}"
+
+        # Block waiting for `sc` to print "you may start your tests"
+        tests_started, log_lines = tests_started?(log_file, pid)
+
+        if tests_started
+          log "success, sc is running"
+          unless daemonize
+            at_exit do
+              Process.kill("TERM", pid)
+            rescue Errno::ESRCH
+            end
+          end
+          return pid
+        else
+          puts
+          puts log_lines.join
+          puts
+          log "ERROR: couldn't start sc"
+        end
+      ensure
+        log_file.close
+      end
+
+      private def open_log_file
+        log_file_path = deploy_dir('log/sc.log')
+        FileUtils.mkdir_p File.dirname(log_file_path) # create log dir if it doesn't exist
+        File.open(log_file_path, 'a+') # open log file, create if it doesn't exist, seek to the end
+      end
+
+      private def process_exited?(pid)
+        Process.waitpid(pid, Process::WNOHANG)
+      end
+
+      # Scan the log output of sc looking for the "you may start your tests" message
+      # timeout if more than SC_START_TIMEOUT_S seconds pass without seeing the message
+      private def tests_started?(log_file, pid)
+        lines = []
+        Timeout.timeout(SC_START_TIMEOUT_S) do
+          # Keep reading from the end of log_file unless the process exits
+          until process_exited? pid
+            sleep 0.1
+            while (line = log_file.gets)
+              lines << line
+              if line.strip.end_with?(SC_START_MESSAGE)
+                return true, lines
+              end
+            end
+          end
+          lines << log_file.readlines # read any remaining lines
+        end
+        return false, lines
+      rescue Timeout::Error
+        log "ERROR: timed out waiting #{SC_START_TIMEOUT_S} seconds for '#{SC_START_MESSAGE}', stopping sc"
+        begin
+          Process.kill("TERM", pid) # stop sc
+        rescue Errno::ESRCH
+        end
```

Comment:
worth pulling this pattern out of both here and the `if tests_started` block into a helper method?

Reply:
👍, I love how in a ruby method you get a `rescue` clause for free without having to start a begin/end, such a nice subtle detail that makes short ruby methods a big win

Reply:
Likewise! I might not have suggested it otherwise